Windows builds failing due to libclang v16

Noticing that libclang 16 was published today sometime between ([old build](https://github.com/bytecodealliance/wasmtime-rb/actions/runs/4579552856/jobs/8087453358) .. [new build](https://github.com/bytecodealliance/wasmtime-rb/actions/runs/4584581454/jobs/8096161127)). I'm not 100% sure yet, but I _think_ this is a bindgen issue that may be fixed on `main`. Unfortunately, `v0.9` version of rb-sys cannot upgrade `bindgen` due to breaking MSRV changes. So we'll have to find another solution.

Awhile ago, I made a `libclang-rb` which packages native v14 binaries of libclang, so theoretically we should be able to hook that up. Something like:

```sh
$ gem install libclang
$ echo "LIBCLANG_PATH=$(ruby -rlibclang -e 'puts Libclang.libdir')" > $GITHUB_ENV
```

On Ruby 3.1, I get a different error when using `/ucrt64/include`:

```
BINDGEN_EXTRA_CLANG_ARGS="-I/ucrt64/include" bundle exec rake

[...]

thread 'main' panicked at 'generate bindings: ClangDiagnostic("C:/msys64/ucrt64/include\\stdlib.h:388:47: error: expected ';' after top level declarator\nC:/msys64/ucrt64/include\\stdlib.h:389:48: error: expected ';' after top level declarator\nC:/msys64/ucrt64/include\\stdlib.h:391:53: error: expected ';' after top level declarator\nC:/msys64/ucrt64/include\\stdlib.h:396:26: error: expected ';' after top level declarator\nC:/msys64/ucrt64/include\\stdlib.h:686:69: error: expected
[...]
```

Not sure if that's helpful.

--- a/setup-ruby-and-rust/action.yml
+++ b/setup-ruby-and-rust/action.yml
@@ -279,9 +279,9 @@ runs:
         msys_root=""
 
         if [ "$MSYSTEM" = "MINGW64" ]; then
-          msys_root="$RI_DEVKIT/mingw64"
+          msys_root="c:/msys64/mingw64"
         elif [ "$MSYSTEM" = "UCRT64" ]; then
-          msys_root="$RI_DEVKIT/ucrt64"
+          msys_root="c:/msys64/ucrt64"
         else
           echo "Unknown MSYSTEM: $MSYSTEM, cannot configure bindgen"
           exit 1

--- a/setup-ruby-and-rust/action.yml
+++ b/setup-ruby-and-rust/action.yml
@@ -271,7 +271,7 @@ runs:
         echo BINDGEN_EXTRA_CLANG_ARGS="-I$((gcm clang).source -replace 'bin\clang.exe', 'include')" >> $env:GITHUB_ENV
 
     - name: Configure bindgen
-      if: contains(steps.derive-toolchain.outputs.toolchain, 'pc-windows-gnu') == false
+      if: contains(steps.derive-toolchain.outputs.toolchain, 'pc-windows-gnu')
       shell: bash
       run: |
         : Configure bindgen for msys2

--- a/setup-ruby-and-rust/action.yml
+++ b/setup-ruby-and-rust/action.yml
@@ -70,6 +70,8 @@ runs:
         working-directory: ${{ inputs.working-directory }}
         cache-version: ${{ inputs.cache-version }}
         rubygems: ${{ inputs.rubygems }}
+        # for now we have to use clang-15 for mingw builds, since our bindgen is too old to work with clang-16
+        mingw: clang-15
 
     - name: Install helpers
       shell: bash
@@ -93,13 +95,6 @@ runs:
         : Derive toolchain
         derived_toolchain="$(ruby --disable-gems $GITHUB_ACTION_PATH/helpers/derive_toolchain.rb ${{ inputs.rustup-toolchain }})"
         echo "toolchain=$derived_toolchain" >> $GITHUB_OUTPUT
-
-    - name: Warn broken mingw toolchain
-      if: contains(steps.derive-toolchain.outputs.toolchain, 'pc-windows-gnu')
-      shell: bash
-      run: |
-        : Warn broken mingw toolchain
-        echo "::warning::There is known issue with the mingw toolchain which may cause your rb-sys CI builds to fail 😢. Consider temporarily disabling Windows builds until https://github.com/oxidize-rb/actions/issues/20 is resolved #helpwanted."
 
     - uses: dtolnay/rust-toolchain@master
       id: rust-toolchain
@@ -272,5 +267,30 @@ runs:
       if: contains(steps.derive-toolchain.outputs.toolchain, 'msvc')
       shell: pwsh
       run: |
-        echo "LIBCLANG_PATH=$((gcm clang).source -replace "clang.exe")" >> $env:GITHUB_ENV
-        echo "BINDGEN_EXTRA_CLANG_ARGS=$((gcm clang).source -replace "bin\clang.exe","include")" >> $env:GITHUB_ENV
+        echo LIBCLANG_PATH="$((gcm clang).source -replace 'clang.exe')" >> $env:GITHUB_ENV
+        echo BINDGEN_EXTRA_CLANG_ARGS="-I$((gcm clang).source -replace 'bin\clang.exe', 'include')" >> $env:GITHUB_ENV
+
+    - name: Configure bindgen
+      if: contains(steps.derive-toolchain.outputs.toolchain, 'pc-windows-gnu') == false
+      shell: bash
+      run: |
+        : Configure bindgen for msys2
+
+        msys_root=""
+
+        if [ "$MSYSTEM" = "MINGW64" ]; then
+          msys_root="$RI_DEVKIT/mingw64"
+        elif [ "$MSYSTEM" = "UCRT64" ]; then
+          msys_root="$RI_DEVKIT/ucrt64"
+        else
+          echo "Unknown MSYSTEM: $MSYSTEM, cannot configure bindgen"
+          exit 1
+        fi
+
+        # We can't use llvm-16 just yet due to backwards compatibility issues, so we have to workaround for now
+        echo MSYS_ROOT="$msys_root" >> $GITHUB_ENV
+        echo LIBCLANG_PATH="$msys_root/opt/llvm-15/bin" >> $GITHUB_ENV
+        echo BINDGEN_EXTRA_CLANG_ARGS="--target=${{ steps.derive-toolchain.outputs.toolchain }} --sysroot=$msys_root" >> $GITHUB_ENV
+
+        echo "::info::Set LIBCLANG_PATH to $LIBCLANG_PATH"
+        echo "::info::Set BINDGEN_EXTRA_CLANG_ARGS to $BINDGEN_EXTRA_CLANG_ARGS"

